docs/architecture/adr-110-e2e-abci++.md

```diff
@@ -0,0 +1,267 @@
+# ADR 82: E2E tests for CometBFT's behaviour in respect to ABCI++.
+
+## Context
+
+We want to test whether CommetBFT respects the ABCI++ grammar. To do this, we need to enhance the e2e tests infrastructure. Specifically, we plan to do three things:
+- Log every CometBFT's ABCI++ request during the execution.
+- Parse the logs post-mortem and extract all ABCI++ requests.
+- Check if the set of observed requests respects the ABCI++ grammar.
+
+
+Issue: [353](https://github.com/cometbft/cometbft/issues/353).
+
+Current version does not support vote extensions. However, this is the next step. 
+
+
+## Decision
+
+### 1) ABCI++ requests logging
+The idea was to do this at the Application side. Every time the Application 
+receives a request, it logs it.
+
+<strong>Implementation</strong>
+
+The rationale behind this part of the implementation was to log the request concisely and use the existing structures as much as possible. 
+
+Whenever an ABCI request is made, the application will create `abci.Request` (`abci` stands for `"github.com/cometbft/cometbft/abci/types"`) and log it.  The example is below.  
+
+```go
+func (app *Application) InitChain(_ context.Context, req *abci.RequestInitChain) (*abci.ResponseInitChain, error) {
+
+	r := &abci.Request{Value: &abci.Request_InitChain{InitChain: &abci.RequestInitChain{}}}
+	app.logRequest(r)
+    
+	...
+}
+```
+Notice here that we create an empty `abci.RequestInitChain` object while we can also use the one passed to the `InitChain` function. The reason behind this is that, at the moment, we do not need specific fields of the request; we just need to be able to extract the information about the request type. For this, an empty object of a particular type is enough. 
+
+`app.logRequest(r)` function is a new function implemented in the same file (`test/e2e/app/app.go`). Its implementation is the following: 
+
+```go
+func (app *Application) logRequest(req *abci.Request) {
+	s, err := GetABCIRequestString(req)
+	if err != nil {
+		panic(err)
+	}
+	app.logger.Debug(s)
+}
+```
+
+`GetABCIRequestString(req)` is a new method that receives a request and returns its string representation. The implementation and tests for this function and the opposite function `GetABCIRequestFromString(req)`
+that returns `abci.Request` from the string are provided in files `test/e2e/app/log.go` and `test/e2e/app/log_test.go`, respectively. To create a string representation of a request, we first marshal the request via `proto.Marshal()` method and then convert received bytes in the string using `base64.StdEncoding.EncodeToString()` method. The code of this method is below: 
+
+```go
+func GetABCIRequestString(req *abci.Request) (string, error) {
+	b, err := proto.Marshal(req)
+	if err != nil {
+		return "", err
+	}
```

Comment:
As this will go into a log line, consider compressing it before converting to base64 (although not sure what the gains would be... maybe later on when we log more info)

Reply:
At the moment, this will not make much of a difference since objects are empty and hence small. However, I can do it if you think we should do it anyway for the future.

Reply:
I did some experiments to evaluate how big is a single log so we can make a decision around this. And I think at the moment there is not much point in compressing since the size of a single log containing abci.Request is at most 24B. 
However, if we decide to log the full request instead of the empty one then maybe it would make sense to do the compressing since when I changed the logging to log the full request I observed that individual request took up to 7KB.